From 00a978fa621e35c773561db86e6f33b506ae7b17 Mon Sep 17 00:00:00 2001
From: Emmanuel Berthier <emmanuel.berthier@intel.com>
Date: Fri, 8 Apr 2016 11:12:14 +0200
Subject: [PATCH 3/5] [MAGN] out of bounds access in ak09911_verify_chip_id

KASAN reports:

[   28.932231] BUG: AddressSanitizer: out of bounds access in memcpy+0x2d/0x50 at addr ffff880064717868
[   28.942471] Write of size 3 by task init/1
[   28.947084] page:ffffea000191c5c0 count:0 mapcount:0 mapping:          (null) index:0x0
[   28.956053] page flags: 0x4000000000000000()
[   28.960891] page dumped because: kasan: bad access detected
[   28.967157] CPU: 0 PID: 1 Comm: init Not tainted 3.14.55-x86_64-00983-g396cea9-dirty #6
[   28.976135] Hardware name: Insyde CherryTrail/T3 MRD, BIOS CHTMRD.A6.001.007 03/23/2016
[   28.985945]  [<ffffffff820304bb>] dump_stack+0x4e/0x7a
[   28.991733]  [<ffffffff8127a148>] kasan_report_error+0x448/0x450
[   28.998484]  [<ffffffff8127947c>] __asan_storeN+0x1ec/0x200
[   29.004756]  [<ffffffff8127992d>] ? memcpy+0x2d/0x50
[   29.010348]  [<ffffffff8127992d>] memcpy+0x2d/0x50
[   29.015746]  [<ffffffff81b616f2>] i2c_smbus_read_i2c_block_data+0x102/0x150
[   29.023568]  [<ffffffff81b615f0>] ? i2c_smbus_write_block_data+0x130/0x130
[   29.031297]  [<ffffffff82042d18>] ? _raw_spin_unlock_irqrestore+0x28/0x60
[   29.038925]  [<ffffffff81128b66>] ? up+0x46/0x60
[   29.044127]  [<ffffffff8112b25a>] ? do_raw_spin_lock+0xea/0x210
[   29.050797]  [<ffffffffa00180c0>] ? ak09911_set_mode+0xa0/0xa0 [ak09911]
[   29.058334]  [<ffffffffa0018137>] ak09911_probe+0x77/0x440 [ak09911]

Change-Id: I50ddf77ee73d99481c07eb9f668200a811d161aa
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-28358
Signed-off-by: Emmanuel Berthier <emmanuel.berthier@intel.com>
Reviewed-on: https://android.intel.com:443/492321
---
 drivers/iio/magnetometer/ak09911.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/iio/magnetometer/ak09911.c b/drivers/iio/magnetometer/ak09911.c
index 4e36e1d..f054303 100644
--- a/drivers/iio/magnetometer/ak09911.c
+++ b/drivers/iio/magnetometer/ak09911.c
@@ -178,7 +178,7 @@ static int ak09911_verify_chip_id(struct i2c_client *client)
 	int ret;
 
 	ret = i2c_smbus_read_i2c_block_data(client, AK09911_REG_WIA1,
-					    3, wia_val);
+					    2, wia_val);
 	if (ret < 0) {
 		dev_err(&client->dev, "Error reading WIA\n");
 		return ret;
-- 
1.9.1

